xarray/backends/rasterio_.py

```diff
@@ -268,9 +274,13 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None,
         attrs['is_tiled'] = np.uint8(riods.is_tiled)
     if hasattr(riods, 'nodatavals'):
         # The nodata values for the raster bands
-        attrs['nodatavals'] = tuple(
+        nodatavals = tuple(
             np.nan if nodataval is None else nodataval
             for nodataval in riods.nodatavals)
+        if mask:
+            encoding['nodatavals'] = nodatavals
```

Comment:
I thing setting the encoding should inline with the other if statements and should be something like:
```python
if mask and riods.nodata is not None:
    encoding['_FillValue'] = riods.nodata
```

Reply:
@shoyer is this what you intended or did you want to keep it as @jsignell modified it to be?